hazelcast/src/main/java/com/hazelcast/internal/util/ConcurrencyDetection.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import static java.lang.System.nanoTime;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * Responsible for tracking if concurrency is detected.
+ * <p>
+ * Concurrency is currently triggered at 2 levels:
+ * - number of concurrent invocations
+ * - contention on the io system.
+ */
+public abstract class ConcurrencyDetection {
+
+    private final boolean enabled;
+
+    private ConcurrencyDetection(boolean enabled) {
+        this.enabled = enabled;
+    }
+
+    /**
+     * Checks if any concurrency is detected.
+     * <p>
+     * This call is thread-safe;
+     *
+     * @return true if detected, false otherwise.
+     */
+    public abstract boolean isDetected();
+
+    /**
+     * Called when concurrency is detected.
+     * <p>
+     * This call is thread-safe;
+     */
+    public abstract void onDetected();
+
+    /**
+     * Checks if this ConcurrencyDetection is enabled.
+     * <p>
+     * This is useful if you want to fall back to legacy mode.
+     *
+     * @return true if enabled, false otherwise.
+     */
+    public boolean enabled() {
+        return enabled;
+    }
+
+    public static ConcurrencyDetection createEnabled(long durationMs) {
+        return new EnabledConcurrencyDetection(durationMs);
+    }
+
+    public static ConcurrencyDetection createDisabled() {
+        return new DisabledConcurrencyDetection();
+    }
+
+    /**
+     * The always AlwaysConcurrencyDetection indicates that there is always concurrency,
+     * even if there is on concurrency. This prevent write through and therefor lets the
+     * system behave as before the write through was added.
+     */
+    private static final class DisabledConcurrencyDetection extends ConcurrencyDetection {
+
+        public DisabledConcurrencyDetection() {
+            super(false);
+        }
+
+        @Override
+        public boolean isDetected() {
+            return true;
+        }
+
+        @Override
+        public void onDetected() {
+            //no-op
+        }
+    }
+
+    private final static class EnabledConcurrencyDetection extends ConcurrencyDetection {
```

Comment:
Can you explain the algorithm here? How concurrency detection works? How it's notified? How fast it reacts to the system load? etc...

Reply:
There are 2 sources of detecting concurrency:
- the call id generator that will track the number of inflight calls. If the number of inflight calls exceeds a certain threshold, the 'onDetected' is called.
- the outbound pipeline. So we have a scheduled flag and if you don't manage to cas it, it means there is concurrency and we call 'onDetected'.

The concurrency detection uses a time window. As long as concurrency is detected, the window keeps sliding forward. But if no concurrency is detected, the window will eventually be smaller than the current time and then no concurrency is detected.

Reply:
I see. So once concurrency is detected, it will be effective immediately. But even concurrent access stops, it will report as _concurrency detected_ until window timeouts.

Reply:
Can you add that explanation to the javadoc of the class?

Reply:
I guess the one more important detail is this:
```
if (nowNanos - (expirationNanos - halfWindowNanos) > 0) {
                expirationNanosRef.lazySet(nowNanos + windowNanos);
            }
```
It means if only half-window time has passed since the last time we detected concurrency, it will not be prolonged. For instance, concurrency is detected at 500 and lasts until 600. It's detected again at 540 but it still lasts until 600. It's detected at 560 and now prolonged until 660. Is this right?

Reply:
correct

Reply:
The idea is to prevent contention on the cache line when there is concurrency. So instead of always kicking the window forward, you only kicked it once and a while.